internal/controller/deployment_to_pdb_controller_test.go

```diff
@@ -0,0 +1,181 @@
+		// Create the Namespace object (from corev1)
+		namespaceObj := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespace,
+			},
+		}
+
+		// create the namespace using the controller-runtime client
+		_ = k8sClient.Create(context.Background(), namespaceObj)
+
+		// Create a fake clientset and add required schemas
+		s := scheme.Scheme
+		Expect(appsv1.AddToScheme(s)).To(Succeed())
+		Expect(policyv1.AddToScheme(s)).To(Succeed())
+
+		surge := intstr.FromInt(1)
+		// Create the reconciler instance
+		r = &DeploymentToPDBReconciler{
+			Client: k8sClient, // Use the fake client
+			Scheme: s,
+		}
+
+		// Define a Deployment to test
+		deployment = &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      deploymentName,
+				Namespace: namespace,
+			},
+			Spec: appsv1.DeploymentSpec{
+				Replicas: int32Ptr(3),
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"app": "example",
+					},
+				},
+				Strategy: appsv1.DeploymentStrategy{
+					RollingUpdate: &appsv1.RollingUpdateDeployment{
+						MaxSurge: &surge,
+					},
+				},
+				Template: corev1.PodTemplateSpec{ // Use corev1.PodTemplateSpec
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{
+							"app": "example",
+						},
+					},
+					Spec: corev1.PodSpec{ // Use corev1.PodSpec
+						Containers: []corev1.Container{ // Use corev1.Container
+							{
+								Name:  "nginx",
+								Image: "nginx:latest",
+							},
+						},
+					},
+				},
+			},
+		}
+
+		// Create the deployment
+		_ = r.Client.Create(context.Background(), deployment)
+		//Expect(err).To(BeNil())
+	})
+
+	AfterEach(func() {
+		// Create the PDB with a deletion timestamp set
+		pdb := &policyv1.PodDisruptionBudget{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      deploymentName + "-pdb",
+				Namespace: namespace,
+			},
+		}
+		_ = r.Client.Delete(context.Background(), pdb)
+		//Expect(err).To(BeNil())
+
+	})
+
+	Describe("when a deployment is created", func() {
+		It("should create a PodDisruptionBudget", func() {
+			req := reconcile.Request{
+				NamespacedName: client.ObjectKey{
+					Namespace: namespace,
+					Name:      deploymentName,
+				},
+			}
+
+			// Call the reconciler
+			_, err := r.Reconcile(context.Background(), req)
+			Expect(err).To(BeNil())
+
+			// Check if PDB is created
+			pdb := &policyv1.PodDisruptionBudget{}
+			err = r.Client.Get(context.Background(), client.ObjectKey{
+				Namespace: namespace,
+				Name:      deploymentName + "-pdb",
+			}, pdb)
+			Expect(err).To(BeNil())
+
+			Expect(pdb.Name).To(Equal(deploymentName + "-pdb"))
+			Expect(pdb.Spec.MinAvailable.IntVal).To(Equal(int32(3)))
+		})
+	})
+
+	Describe("when a deployment is deleted", func() {
+		It("should delete the corresponding PodDisruptionBudget", func() {
+
+			// Now delete the deployment
+			err := r.Client.Delete(context.Background(), deployment)
+			Expect(err).To(BeNil())
+
+			req := reconcile.Request{
```

Comment:
so we might need a finalizer/deletion timetsamtmp to do this correctly. Lets say this controller is temporarily down and it misses a delete. How will it find it later thats what finalizers were made for.  Or we just accept there is some possibility of a leak and move on (if deployment comes back with different replica count then we'll just scale once thats handled)

Reply:
anyways might be worth creating an issue saying we can leak here even if it's not a big deal.

Reply:
Ironcally leaking pdbs would block our current upgrade logic we had to toggle off wher  expected pods  == 0